File: src-tauri/src/todo.rs
// ──────────────────────────────────────────────────────────
// Todo — markdown-backed list at .0canvas/todo.md
// ──────────────────────────────────────────────────────────
//
// Intentionally minimal. The source of truth is a plain
// markdown file the user (or an agent, once Phase 4 lands)
// can also edit directly. We parse `- [ ]` / `- [x]` lines
// as actionable items; every other line is preserved verbatim
// on round-trip so headings, blank lines, and notes survive.
// ──────────────────────────────────────────────────────────

use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;

use crate::sidecar::SidecarState;

#[derive(Serialize, Deserialize, Clone)]
pub struct TodoItem {
    /// 0-based index of the line in the source file.
    pub line: usize,
    /// True for `- [x]`, false for `- [ ]`.
    pub done: bool,
    pub text: String,
}

#[derive(Serialize, Deserialize, Clone)]
pub struct TodoFile {
    /// Absolute path to `.0canvas/todo.md`.
    pub path: String,
    pub raw: String,
    pub items: Vec<TodoItem>,
}

fn todo_path(root: &PathBuf) -> PathBuf {
    root.join(".0canvas").join("todo.md")
}

/// Very simple checkbox-line parser. Accepts:
///   "- [ ] task"   → { done:false, text:"task" }
///   "- [x] task"   → { done:true,  text:"task" }
///   "* [X] task"   → same as above (case-insensitive x, allows *)
/// Leading whitespace is preserved as-is when writing back but NOT
/// shown in the text field — the UI would otherwise drift indent on
/// every edit.
fn parse_todo_line(line: &str, idx: usize) -> Option<TodoItem> {
    let trimmed = line.trim_start();
    let bytes = trimmed.as_bytes();
    if bytes.len() < 6 {
        return None;
    }
    // `- ` or `* `
    let (_marker, rest) = if trimmed.starts_with("- ") {
        ("-", &trimmed[2..])
    } else if trimmed.starts_with("* ") {
        ("*", &trimmed[2..])
    } else {
        return None;
    };
    if !rest.starts_with('[') {
        return None;
    }
    // `[ ]` or `[x]` or `[X]`
    let inner = rest.chars().nth(1)?;
    let done = match inner {
        ' ' => false,
        'x' | 'X' => true,
        _ => return None,
    };
    if rest.chars().nth(2) != Some(']') {
        return None;
    }
    let after = rest[3..].trim_start();
    Some(TodoItem {
        line: idx,
        done,
        text: after.to_string(),
    })
}

#[tauri::command]
pub fn load_todo_file(state: tauri::State<'_, SidecarState>) -> Result<TodoFile, String> {
    let root = state
        .current_root()
        .ok_or_else(|| "no project root".to_string())?;
    let path = todo_path(&root);

    let raw = match fs::read_to_string(&path) {
        Ok(s) => s,
        Err(_) => String::new(), // file doesn't exist yet; treat as empty
    };
    let items: Vec<TodoItem> = raw
        .lines()
        .enumerate()
        .filter_map(|(i, line)| parse_todo_line(line, i))
        .collect();

    Ok(TodoFile {
        path: path.to_string_lossy().into_owned(),
        raw,
        items,
    })
}

/// Write the full markdown back. The frontend sends the complete raw
/// content so we don't have to guess about whitespace / comments.
#[tauri::command]
pub fn save_todo_file(state: tauri::State<'_, SidecarState>, raw: String) -> Result<(), String> {
    let root = state
        .current_root()
        .ok_or_else(|| "no project root".to_string())?;
    let path = todo_path(&root);

    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|e| format!("create_dir_all: {}", e))?;
    }

    let mut content = raw;
    if !content.ends_with('\n') {
        content.push('\n');
    }

    let tmp = path.with_extension("md.0canvas-tmp");
    fs::write(&tmp, &content).map_err(|e| format!("write tmp: {}", e))?;
    fs::rename(&tmp, &path).map_err(|e| format!("rename: {}", e))?;
    Ok(())
}

File: electron/ipc/commands/todo.ts
// ──────────────────────────────────────────────────────────
// IPC commands: todo
// ──────────────────────────────────────────────────────────
//
// Markdown-backed todo list at <project>/.zeros/todo.md. Checkbox
// lines (`- [ ]` / `- [x]`, or `*` variants) are actionable items;
// every other line is preserved verbatim on round-trip so headings,
// blanks, and notes survive editing.
// ──────────────────────────────────────────────────────────

import fs from "node:fs";
import path from "node:path";
import { currentRoot } from "../../sidecar";
import type { CommandHandler } from "../router";

interface TodoItemPayload {
  line: number;
  done: boolean;
  text: string;
}

interface TodoFilePayload {
  path: string;
  raw: string;
  items: TodoItemPayload[];
}

function resolveRoot(args: Record<string, unknown>): string {
  const explicit = typeof args.cwd === "string" ? args.cwd.trim() : "";
  if (explicit) return explicit;
  const fallback = currentRoot();
  if (!fallback) throw new Error("no project root");
  return fallback;
}

function todoPath(root: string): string {
  return path.join(root, ".zeros", "todo.md");
}

/** Parse one checkbox line. Matches parse_todo_line() in todo.rs:
 *    "- [ ] task"  → { done:false, text:"task" }
 *    "- [x] task"  → { done:true,  text:"task" }
 *    "* [X] task"  → same as above (case-insensitive, allows * marker)
 *  Leading whitespace on the source line is NOT returned in `text`
 *  so the UI doesn't drift indentation on every round-trip. */
function parseTodoLine(line: string, idx: number): TodoItemPayload | null {
  const trimmed = line.replace(/^\s+/, "");
  if (trimmed.length < 6) return null;

  let rest: string;
  if (trimmed.startsWith("- ")) rest = trimmed.slice(2);
  else if (trimmed.startsWith("* ")) rest = trimmed.slice(2);
  else return null;

  if (!rest.startsWith("[")) return null;

  const inner = rest.charAt(1);
  let done: boolean;
  if (inner === " ") done = false;
  else if (inner === "x" || inner === "X") done = true;
  else return null;

  if (rest.charAt(2) !== "]") return null;

  const after = rest.slice(3).replace(/^\s+/, "");
  return { line: idx, done, text: after };
}

export const loadTodoFile: CommandHandler = (args) => {
  const root = resolveRoot(args);
  const p = todoPath(root);
  let raw = "";
  try {
    raw = fs.readFileSync(p, "utf-8");
  } catch {
    // file may not exist yet -> treat as empty
  }
  const items: TodoItemPayload[] = [];
  // `.split(/\r?\n/)` would keep a final empty row; `.trimEnd()` avoids that.
  const lines = raw.split("\n");
  const hasTrailingNewline = raw.endsWith("\n");
  const iterLen = hasTrailingNewline ? lines.length - 1 : lines.length;
  for (let i = 0; i < iterLen; i++) {
    const item = parseTodoLine(lines[i], i);
    if (item) items.push(item);
  }
  const payload: TodoFilePayload = { path: p, raw, items };
  return payload;
};

export const saveTodoFile: CommandHandler = (args) => {
  const root = resolveRoot(args);
  const p = todoPath(root);
  const raw = String(args.raw ?? "");

  const parent = path.dirname(p);
  try {
    fs.mkdirSync(parent, { recursive: true });
  } catch (err) {
    throw new Error(
      `create_dir_all: ${err instanceof Error ? err.message : String(err)}`,
    );
  }

  let content = raw;
  if (!content.endsWith("\n")) content += "\n";

  const tmp = p.replace(/\.md$/, ".md.zeros-tmp");
  try {
    fs.writeFileSync(tmp, content);
  } catch (err) {
    throw new Error(
      `write tmp: ${err instanceof Error ? err.message : String(err)}`,
    );
  }
  try {
    fs.renameSync(tmp, p);
  } catch (err) {
    throw new Error(
      `rename: ${err instanceof Error ? err.message : String(err)}`,
    );
  }
};
